transport/tlsfrag/buffer.go

```diff
@@ -0,0 +1,134 @@
+}
+
+var _ io.Writer = (*clientHelloBuffer)(nil)
+var _ io.ReaderFrom = (*clientHelloBuffer)(nil)
+
+// newClientHelloBuffer creates and initializes a new buffer to receive a TLS Client Hello packet.
+func newClientHelloBuffer() *clientHelloBuffer {
+	// Allocate the 5 bytes header first, and then reallocate it to contain the entire packet later
+	return &clientHelloBuffer{
+		data:    make([]byte, 0, recordHeaderLen),
+		padding: 0,
+		valid:   true,
+		bufrd:   bytes.NewReader(nil), // It will be Reset in Write
+	}
+}
+
+// Bytes returns the full Client Hello packet including both the 5 bytes header and the content.
+func (b *clientHelloBuffer) Bytes() []byte {
+	return b.data
+}
+
+// Write appends p to the buffer and returns the number of bytes actually used.
+// If this data completes a valid TLS Client Hello, it returns errTLSClientHelloFullyReceived.
+// If an invalid TLS Client Hello message is detected, it returns the error errInvalidTLSClientHello.
+// If all bytes in p have been used and the buffer still requires more data to build a complete TLS Client Hello
+// message, it returns (len(p), nil).
+func (b *clientHelloBuffer) Write(p []byte) (int, error) {
+	b.bufrd.Reset(p)
+	n, err := b.ReadFrom(b.bufrd)
+	if err == nil && int(n) != len(p) {
+		err = io.ErrShortWrite
+	}
+	return int(n), err
+}
+
+// ReadFrom reads all the data from r and appends it to this buffer until a complete Client Hello packet has been
+// received, or r returns EOF or error. It returns the number of bytes read. Any error except EOF encountered during
+// the read is also returned.
+//
+// If this buffer completes a valid TLS Client Hello, it returns errTLSClientHelloFullyReceived.
+// If an invalid TLS Client Hello message is detected, it returns the error errInvalidTLSClientHello.
+// If this buffer still requires more data to build a complete TLS Client Hello message, it returns nil error.
+//
+// You can call ReadFrom multiple times if r doesn't provide enough data to build a complete Client Hello packet.
+func (b *clientHelloBuffer) ReadFrom(r io.Reader) (n int64, err error) {
+	if !b.valid {
+		return 0, errInvalidTLSClientHello
+	}
+
+	for len(b.data) < cap(b.data)-b.padding && err == nil {
+		m, e := r.Read(b.data[len(b.data) : cap(b.data)-b.padding])
+		b.data = b.data[:len(b.data)+m]
+		n += int64(m)
+		err = e
+
+		if len(b.data) == recordHeaderLen {
+			hdr, e := b.validateTLSClientHello()
+			if e != nil {
+				if err == io.EOF {
+					err = nil
+				}
+				err = errors.Join(err, e)
+				return
+			}
+			buf := make([]byte, 0, recordHeaderLen*2+hdr.PayloadLen())
+			b.data = append(buf, b.data...)
+			b.padding = recordHeaderLen
+		}
+	}
+
+	if err == io.EOF {
+		err = nil
+	}
+	if len(b.data) == cap(b.data)-b.padding {
+		err = errors.Join(err, errTLSClientHelloFullyReceived)
+	}
+	return
+}
+
+func (b *clientHelloBuffer) validateTLSClientHello() (tlsRecordHeader, error) {
```

Comment:
Can you make this a standalone function by passing `b.data` as input? Clear inputs are a lot easier to understand. Otherwise it's hard to tell it doesn't use any other data.
You can set `b.valid` in the caller.